lib/packageModules.js

```diff
@@ -26,6 +26,73 @@ function setArtifactPath(funcName, func, artifactPath) {
   }
 }
 
+function serverlessZip(args) {
+  const artifactFilePath = args.artifactFilePath;
+  const directory = args.directory;
+  const files = args.files;
+
+  const zip = archiver.create('zip');
+  const output = fs.createWriteStream(artifactFilePath);
+  return new BbPromise((resolve, reject) => {
+    output.on('close', () => resolve(artifactFilePath));
+    output.on('error', err => reject(err));
+    zip.on('error', err => reject(err));
+
+    output.on('open', () => {
+      zip.pipe(output);
+
+      // normalize both maps to avoid problems with e.g. Path Separators in different shells
+      const normalizedFiles = _.uniq(_.map(files, file => path.normalize(file)));
+
+      BbPromise.all(_.map(normalizedFiles, file => getFileContentAndStat.call(this, directory, file)))
+        .then(contents => {
+          _.forEach(
+            contents.sort((content1, content2) => content1.filePath.localeCompare(content2.filePath)),
+            file => {
+              const name = file.filePath;
+              // Ensure file is executable if it is locally executable or
+              // we force it to be executable if platform is windows
+              const mode = file.stat.mode & 0o100 || process.platform === 'win32' ? 0o755 : 0o644;
+              zip.append(file.data, {
+                name,
+                mode,
+                date: new Date(0) // necessary to get the same hash when zipping the same content
+              });
+            }
+          );
+
+          return zip.finalize();
+        })
```

Comment:
Both `date: new Date(0)` and `contents.sort` are necessary to get the same hash. I'd mention that in the single block comment, for the future generations :wink:

Reply:
> Both `date: new Date(0)` and `contents.sort` are necessary to get the same hash. I'd mention that in the single block comment, for the future generations ![😉](chrome-extension://cahedbegdkagmcjfolhdlechbkeaieki/images/apple/1f609.png)

I could, but my intent was to copy serverless' implementation directly (this is copy-pasted). I don't think it's a huge risk but if anything, making changes would make it more difficult to track changes against serverless' (not that I would personally do this work, but someone might want to)